source/Esf.Website/App/esfState/esfStateSaveCommand.service.ts

```diff
@@ -0,0 +1,57 @@
+﻿import { Injectable } from '@angular/core';
+import { ActivatedRoute, Router } from '@angular/router';
+import { EsfStateViewModel } from '../common/models/esfStateViewModel';
+import { EsfStateProvider } from '../common/interfaces/esfStateProvider.interface';
+import { EsfStateService } from './esfState.service';
+import { EsfStateValidationService, EsfStateValidationResult } from '../common/services/esfStateValidation.service';
+import { ExistingEsfStateDto } from '../common/models/existingEsfStateDto';
+import { EsfStateDto } from '../common/models/esfStateDto';
+import { CommandState } from '../common/models/commandState';
+import { Observable } from 'rxjs/Observable';
+import { Subject } from 'rxjs/Subject';
+
+@Injectable()
+export class EsfStateSaveCommandService {
+    private stateProvider: EsfStateProvider;
+    private commandStateStream: Subject<CommandState>;
+
+    constructor(
+        private esfStateService: EsfStateService,
+        private stateValidationService: EsfStateValidationService,
+        private router: Router
+    ) {
+        this.commandStateStream = new Subject<CommandState>();
+    }
+
+    public attachStateProvider(stateProvider: EsfStateProvider): void {
+        this.stateProvider = stateProvider;
+    }
+
+    public getCommandState(): Observable<CommandState> {
+        return this.commandStateStream.asObservable();
+    }
+
+    public run(): void {
```

Comment:
If we use Observerveble and Provide the state and the result in it we can expres those run command more directly. Before running command we can initialize the starting state from each provider by calling getInitialState()